A macro to get the current function name. (#2818)
rust-lang/rfcs: text/0000-fn_name_macro.md

- Feature Name: fn_name_macro
- Start Date: 2019-11-14
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

<!-- https://github.com/rust-lang/rfcs/issues/1743 -->

# Summary
[summary]: #summary

This RFC adds an additional macro, `function!`, to the `core` crate. When invoked, the macro expands to the name of the function that contains the call site.

# Motivation

> [motivation]: #motivation

Reviewer: Some elaboration on the common use case would be good. In the issue you mention:
> I wouldn't normally think to pair func_name!() with line!() and file!() because usually when i want it I want "emergency println debugging" and i'm just throwing in a few printers and i delete them once i've sorted out what's wrong.

and I'm wondering whether you expect the macro to be used in e.g. `dbg!(...)` (which does print out `file!()` and `line!()`).

Author: Good point. I will add an example usage to the guide level explanation.

This is a useful extension of Rust's existing debug reporting: `file!`, `line!`, `column!` and `module_path!`.

> For most people the name of the function is a much more immediate way to understand the context of a message than file name and line number.

Reviewer: ```suggestion
For most people, the name of the function is a much more immediate way to understand the context of a message than file name and line number.
```

> For most people the name of the function is a much more immediate way to understand the context of a message than file name and line number.

Reviewer: Also I'm not sure I agree with this, especially not when the claim is "much more". In particular, the `dbg!()` output is structured in such a way that e.g. in VSCode's integrated terminal, the output becomes a clickable link that immediately takes you to the source location. That seems hard to beat...?

Author: That assumes that you are looking at the message in the VS Code integrated terminal. One of my machines doesn't even run VS Code.

Often I see panic or log messages that are pasted into Discord/Zulip/Riot/Github, where you also cannot click the `file:line` text.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> For debug information about what's happening within the program there's several useful macros that you might use. One of them is the `function!` macro, which expands to the name of the current function. If used outside of a function it causes a compilation error.

Reviewer 2: I think returning an `Option<&'static str>` or some special name like `"(top-level)"` is more useful than a compilation error. It is surprising when some macro cannot be used outside a function.

Reviewer 3: Definitely not an Option, that would make it much more painful to use.

Reviewer: Well if we generalize this to items rather than to functions then there is always an enclosing item wherever this is used except directly in top level modules so erroring becomes much less common.

> For debug information about what's happening within the program there's several useful macros that you might use. One of them is the `function!` macro, which expands to the name of the current function. If used outside of a function it causes a compilation error.

Reviewer: ```suggestion
For debug information about what's happening within the program there are several useful macros that you might use. One of them is the `function!` macro, which expands to the name of the current function. If used outside of a function it causes a compilation error.
```

# Reference-level explanation

> [reference-level-explanation]: #reference-level-explanation

Reviewer: This section needs some elaboration to consider the various cases (e.g. like the one @kennytm raised) and give examples of code and what the result is.

Also, looking at the implementation of `module_path!` in `libsyntax_ext` https://github.com/rust-lang/rust/blob/9e8c4e6fb1c952048fb823e59f4c9c6487bf9a58/src/libsyntax_ext/source_util.rs#L65-L73 it looks like the information is readily available. In this case however, the information is not available in `cx` or reachable fields. Some elaboration on the implementation would be good. cc @petrochenkov since they are the most likely T-compiler reviewer for this RFC. Also, this probably has little to do with "debuginfo" -- that's a different part of the compiler that comes into play much later in the compilation process.

> Use of the `function!` macro expands to the compiler's internal name for the function. This will generally be the name that the user wrote into the file but in the case of a closure or similar it will be something like the function's name with a unique suffix.

Reviewer 4: What about a closure in a static?

```rust
static F: &(dyn Fn() -> &'static str + Sync) = &|| "1";
```

(the fully-qualified name is currently `playground::F::{{closure}}`)

Author: @eddyb know this one (i hope)

> Use of the `function!` macro expands to the compiler's internal name for the function. This will generally be the name that the user wrote into the file but in the case of a closure or similar it will be something like the function's name with a unique suffix.

Reviewer 2: An alternative is to return the name of the syntactic first ancestor function containing the macro expression so
```rust
fn process_items(items: &[Item]) {
    process_items.iter().for_each(|item| {
        dbg_with_function_name!(item);
        process(item);
    });
}
```
prints `[process_items:xxx]: Item { ... }` regardless of whether `for` is used or iterator is used.

Author: I'm going to add centril's "item_path!" idea, which covers this case.

The exact text is considered "debug" information and not subject to Rust's stability guarantee.

# Drawbacks

> [drawbacks]: #drawbacks

Reviewer: It's another blow to incremental compilation (well it doesn't get worse than `column!()`, but this adds more...).

Reviewer 3: Seems much less likely to cause such problems than column.

Reviewer: Yeah (I said as much), but it should still be noted in the text to provide a more full examination of the trade-offs.

Author: How is incremental compilation particularly negatively affected? It seems like this would be a lot _less_ sensitive to file changes than `line!` and `column!` are.

Reviewer: Yeah it is less sensitive than `line!()` certainly. It's probably the case that `function!()` or `item_path!()` is as bad as `module_path!()` is; only if you rename the item or parent module does the macro output change, so there's probably not a lot of impact on incremental after all.

* Doing this adds another macro to `core`.
* This macro in particular cannot be implemented in Rust itself, it requires special support from `rustc`, as well as from all potential alternative compilers for the Rust language.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

Previous discussion was within an issue of the RFCs repo: https://github.com/rust-lang/rfcs/issues/1743

> In summary, this is _extremely_ useful for debug purposes, and has been highly desired for nearly two years.

Reviewer: ```suggestion
In summary, this is useful for debug purposes, and has been desired for nearly two years.
```

Reviewer 5: Another use case I'll be using it for: testing. In many of the tests I've written (and this reflects the kinds of crates I've been writing too), temporary directories and other resources are named according to the name of the test function. Not having to keep them in sync manually would be great.

Reviewer: Not sure why you wrote this here, but your use case sounds like you need guarantees wrt. the output (or else your tests might suddenly start failing on an upgrade) and this RFC explicitly disclaims such guarantees.

Reviewer 5: Maybe I should have put it on the section header line. It's more that Github only allows a single conversation per source line.

I don't need stability guarantees, just uniqueness (it also helps to disambiguate log lines in the face of rust-lang/rust#42474).

The exact text is considered "debug" information and not subject to Rust's stability guarantee.

# Drawbacks
[drawbacks]: #drawbacks

* Doing this adds another macro to `core`.
* This macro in particular cannot be implemented in Rust itself, it requires special support from `rustc`, as well as from all potential alternative compilers for the Rust language.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

Previous discussion was within an issue of the RFCs repo: https://github.com/rust-lang/rfcs/issues/1743

In summary, this is useful for debug purposes, and has been desired for nearly two years.

Alternative: we could call it `fn_name!` instead.

Alternative (centril): We could introduce a much more general `item_path!()` macro that is useful in non-function contexts.

```rust
mod foo { // item_path!() ==> foo
    const X: &str = item_path!(); // foo::X
    struct A;
    impl A {
        const Y: &str = item_path!(); // foo::A::Y
        fn fun() {
            // item_path!() ==> foo::A::fun
        }
    }
     type T = Stuff<{item_path!()}> // Stuff<"foo::T">
}
```

# Prior art
[prior-art]: #prior-art

> * C99 has a `__func__` pre-processor macro that expands to the current funciton name. [link](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2004/n1642.html)

Reviewer 6: (Luckily) This is not a pre-processor macro, just a predefined identifier provided by the C compiler
